File: src/main/java/tech/sirwellington/alchemy/generator/NetworkGenerators.java
package tech.sirwellington.alchemy.generator;

import java.net.MalformedURLException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.text.MessageFormat;
import java.util.Arrays;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import tech.sirwellington.alchemy.annotations.access.NonInstantiable;
import tech.sirwellington.alchemy.annotations.arguments.NonEmpty;

import static tech.sirwellington.alchemy.generator.AlchemyGenerator.Get.one;
import static tech.sirwellington.alchemy.generator.Checks.checkNotEmpty;
import static tech.sirwellington.alchemy.generator.Checks.checkThat;
import static tech.sirwellington.alchemy.generator.NumberGenerators.integers;
import static tech.sirwellington.alchemy.generator.PeopleGenerators.popularEmailDomains;
import static tech.sirwellington.alchemy.generator.StringGenerators.alphanumericStrings;

/**
 * {@summary Generators for Network-data, such as an IP Addresses.}
 *
 * @author SirWellington
 */
@NonInstantiable
public final class NetworkGenerators {
    private static final Logger LOG = LoggerFactory.getLogger(NetworkGenerators.class);
    private static final URL FALLBACK_URL;
    private static final List<String> VALID_PROTOCOLS = Arrays.asList(
        "http",
        "https",
        "ftp",
        "file",
        "ssh"
    );

    static {
        URL url;
        try {
            url = new URI("https://google.com").toURL();
        } catch (URISyntaxException | MalformedURLException _) {
            url = null;
        }
        FALLBACK_URL = url;
    }

    private NetworkGenerators() throws IllegalAccessException {
        throw new IllegalAccessException("cannot directly instantiate");
    }

    /**
     * @return {@link URL URLs} beginning with {@code http://}.
     */
    public static AlchemyGenerator<URL> httpURLs() {
        return urlsWithProtocol("http");
    }

    /**
     * @return {@link URL URLs} beginning with {@code https://}.
     */
    public static AlchemyGenerator<URL> httpsURLs() {
        return urlsWithProtocol("https");
    }

    /**
     * Creates valid URLs starting with the provided protocol.
     *
     * @param protocol The protocol to use for the URLs created. Do not include the "://".
     *                 Must be one of the common types, {@code http, https, ftp, ssh, file}.
     * @return {@link URL URLs} beginning with the {@code protocol}.
     */
    public static AlchemyGenerator<URL> urlsWithProtocol(@NonEmpty String protocol) {
        checkNotEmpty(protocol, "missing protocol");
        checkThat(
            VALID_PROTOCOLS.contains(protocol),
            MessageFormat.format("{0} is not a valid protocol [{1}]", protocol, VALID_PROTOCOLS)
        );

        var cleanProtocol = protocol.replace("://", "");
        try {
            new URI(cleanProtocol + "://example.com");
        } catch (URISyntaxException  ex) {
            throw new IllegalArgumentException("Unknown protocol: " + protocol);
        }

        return () -> {
            var hostLength = one(integers(3, 40));
            var host = alphanumericStrings(hostLength);
            var domain = one(popularEmailDomains());
            var url = MessageFormat.format("{0}://{1}.{2}", cleanProtocol, host, domain);

            try {
                return new URI(url).toURL();
            } catch (URISyntaxException | MalformedURLException ex) {
                LOG.error("Failed to create a url from scheme {}", cleanProtocol, ex);
                return FALLBACK_URL;
            }
        };
    }

    /**
     * @return Ports from 22 to {@link Short#MAX_VALUE 32767}.
     */
    public static AlchemyGenerator<Integer> ports() {
        return integers(22, Short.MAX_VALUE);
    }

    /**
     * Generates an IPv4 Address in the form {@code xxx.xxx.xxx.xxx}
     */
    public static AlchemyGenerator<String> ipv4Addresses() {
        var integers = integers(1, 1000);
        return () -> MessageFormat.format(
            "{0}.{1}.{2}.{3}",
            integers.get(),
            integers.get(),
            integers.get(),
            integers.get()
        );
    }
}

File: src/main/java/tech/sirwellington/alchemy/generator/NetworkGenerators.kt
/*
 * Copyright 2017 SirWellington Tech.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


package tech.sirwellington.alchemy.generator


import org.slf4j.LoggerFactory
import tech.sirwellington.alchemy.annotations.access.NonInstantiable
import tech.sirwellington.alchemy.annotations.arguments.NonEmpty
import tech.sirwellington.alchemy.generator.NumberGenerators.Companion.integers
import tech.sirwellington.alchemy.generator.PeopleGenerators.Companion.popularEmailDomains
import tech.sirwellington.alchemy.generator.StringGenerators.Companion.alphanumericString
import java.net.MalformedURLException
import java.net.URL
import java.util.*

/**

 * @author SirWellington
 */
@NonInstantiable
class NetworkGenerators
@Throws(IllegalAccessException::class)
private constructor()
{

    init
    {
        throw IllegalAccessException("cannot instantiate")
    }

    companion object
    {
        private val LOG = LoggerFactory.getLogger(NetworkGenerators::class.java)

        private val FALLBACK_URL: URL

        init
        {
            try
            {
                FALLBACK_URL = URL("http://google.com")
            }
            catch (ex: MalformedURLException)
            {
                throw RuntimeException("Could not generate URL", ex)
            }

        }

        private val VALID_PROTOCOLS = Arrays.asList("http", "https", "ftp", "file", "ssh")

        /**
         * @return URLs beginning with `http://`
         */
        @JvmStatic
        fun httpUrls(): AlchemyGenerator<URL>
        {
            return urlsWithProtocol("http")
        }

        /**
         *
         * @return URLs beginning with `https://`
         */
        @JvmStatic
        fun httpsUrls(): AlchemyGenerator<URL>
        {
            return urlsWithProtocol("https")
        }

        /**
         *
         * @param protocol The protocol to use for the URLs created. Do not include the "://".
         *
         * @return  URLs beginning with the `protocol`
         */
        @JvmStatic
        fun urlsWithProtocol(@NonEmpty protocol: String): AlchemyGenerator<URL>
        {
            checkNotEmpty(protocol, "missing protocol")
            val cleanProtocol = protocol.replace("://", "")

            try
            {
                URL(cleanProtocol + "://")
            }
            catch (ex: MalformedURLException)
            {
                throw IllegalArgumentException("Unknown protocol: " + protocol, ex)
            }

            return AlchemyGenerator {

                val url = "$cleanProtocol://${alphanumericString().get()}.${popularEmailDomains().get()}"

                try
                {
                    URL(url)
                }
                catch (ex: MalformedURLException)
                {
                    LOG.error("Failed to create url from scheme {}", cleanProtocol, ex)
                    FALLBACK_URL
                }
            }
        }

        /**
         * @return Ports from 22 to [32767][Short.MAX_VALUE].
         */
        @JvmStatic
        fun ports(): AlchemyGenerator<Int>
        {
            return integers(22, Short.MAX_VALUE.toInt())
        }

        /**
         * Generates IPV4 addresses in the form `xxx.xx.xxx.xx`.
         * @return An IPV4 address
         */
        @JvmStatic
        fun ip4Addresses(): AlchemyGenerator<String>
        {
            val integers = integers(1, 1000)

            return AlchemyGenerator { "${integers.get()}.${integers.get()}.${integers.get()}.${integers.get()}" }
        }
    }
}
